src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -221,27 +340,98 @@ static ucs_status_t uct_cuda_copy_mem_alloc(uct_md_h md, size_t *length_p,
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemAlloc((CUdeviceptr*)address_p,
-                                                     *length_p));
+
+        for (method = methods; method < methods + num_methods; ++method) {
+            switch (*method) {
+                case UCT_CUDA_ALLOC_METHOD_FABRIC_PINNED:
+                    status = uct_cuda_copy_mem_alloc_fabric(md, alloc_handle);
+                    if (status != UCS_OK) {
+                        continue;
+                    }
+                    goto allocated;
+                case UCT_CUDA_ALLOC_METHOD_LEGACY_PINNED:
+                    status =
+                        UCT_CUDADRV_FUNC_LOG_ERR(cuMemAlloc(&alloc_handle->ptr,
+                                    alloc_handle->length));
+                    if (status != UCS_OK) {
+                        continue;
+                    }
+                    alloc_handle->alloc_type = *method;
+                    goto allocated;
+                default:
+                    status = UCS_ERR_UNSUPPORTED;
+                    goto err;
+            }
+        }
     } else {
         status =
-            UCT_CUDADRV_FUNC_LOG_ERR(cuMemAllocManaged((CUdeviceptr*)address_p,
-                                                       *length_p,
+            UCT_CUDADRV_FUNC_LOG_ERR(cuMemAllocManaged(&alloc_handle->ptr,
+                                                       alloc_handle->length,
                                                        CU_MEM_ATTACH_GLOBAL));
+        alloc_handle->alloc_type = UCT_CUDA_ALLOC_METHOD_LEGACY_MANAGED;
     }
 
+err:
     if (status != UCS_OK) {
+        ucs_free(alloc_handle);
         return status;
     }
 
-    *memh_p = *address_p;
+allocated:
+    *memh_p    = alloc_handle;
+    *address_p = (void*)alloc_handle->ptr;
+    *length_p  = alloc_handle->length;
     return UCS_OK;
 }
 
+static ucs_status_t
+uct_cuda_copy_mem_release_fabric(uct_cuda_copy_alloc_handle_t *alloc_handle)
+{
+    ucs_status_t status;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemRelease(alloc_handle->generic_handle));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemUnmap(alloc_handle->ptr,
+                alloc_handle->length));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemAddressFree(alloc_handle->ptr,
+                alloc_handle->length));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return status;
+}
+
 static ucs_status_t uct_cuda_copy_mem_free(uct_md_h md, uct_mem_h memh)
 {
-    return UCT_CUDADRV_FUNC_LOG_ERR(cuMemFree((CUdeviceptr)memh));
+    uct_cuda_copy_alloc_handle_t *alloc_handle = (uct_cuda_copy_alloc_handle_t *)memh;
+    ucs_status_t status;
+
+#if HAVE_CUDA_FABRIC
+    if (alloc_handle->alloc_type == UCT_CUDA_ALLOC_METHOD_FABRIC_PINNED) {
+        status = uct_cuda_copy_mem_release_fabric(alloc_handle);
+    } else
+#endif
```

Comment:
can we move the `#if` inside uct_cuda_copy_mem_release_fabric? and add { } around the "else" block